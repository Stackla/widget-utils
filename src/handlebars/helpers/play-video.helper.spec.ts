import Handlebars from "handlebars"
import { loadPlayVideoHelper } from "./play-video.helper"

describe("loadPlayVideoHelper", () => {
  beforeAll(() => {
    loadPlayVideoHelper(Handlebars)
  })

  it("should generate a video tag with url, width and height", () => {
    const template = Handlebars.compile(`{{playVideo tile width height}}`)

    const result = template({
      tile: {
        source: "instagram",
        video: {
          standard_resolution: {
            url: "https://example.com/video.mp4"
          }
        }
      },
      width: "100%",
      height: "100%"
    })
    expect(result.trim()).toMatchSnapshot()
  })

  it("should escape the URL, width and height to prevent XSS", () => {
    const template = Handlebars.compile(`{{playVideo tile width}}`)

    const result = template({
      tile: {
        source: "instagram",
        video: {
          standard_resolution: {
            url: "https://example.com/<script>alert(1)</script>.mp4"
          }
        }
      },
      width: '100" onClick="alert(1)',
      height: '100" onClick="alert(1)'
    })

    expect(result.trim()).toMatchSnapshot()
  })

  it("should handle undefined width and height gracefully", () => {
    const template = Handlebars.compile(`{{playVideo tile}}`)

    const result = template({
      tile: {
        source: "instagram",
        video: {
          standard_resolution: {
            url: "https://example.com/video.mp4"
          }
        }
      }
    })
    expect(result.trim()).toMatchSnapshot()
  })

  it("should return an empty string if URL is null or undefined", () => {
    const template = Handlebars.compile(`{{playVideo tile}}`)

    const result = template({
      tile: {
        source: "instagram",
        video: {
          standard_resolution: {
            url: null
          }
        }
      }
    })
    expect(result.trim()).toMatchSnapshot()
  })

  it("should generate a video with embedly html", () => {
    const template = Handlebars.compile(`{{playVideo tile width height}}`)

    const result = template({
      tile: {
        source: "tiktok",
        original_url:
          "https://www.tiktok.com/@jacquefromtheblock/video/7226405581324651819?is_from_webapp=1&sender_device=pc&web_id=7204611010241037826"
      },
      width: "100%",
      height: "100%"
    })
    expect(result.trim()).toMatchSnapshot()
  })
})
