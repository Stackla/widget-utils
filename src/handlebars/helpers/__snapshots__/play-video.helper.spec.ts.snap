// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPlayVideoHelper should escape the URL, width and height to prevent XSS 1`] = `
"<div class="icon-section">
        <div class="top-section">
            
            
        </div>
        <div class="center-section">
            
        </div>
        <div class="bottom-section">
            <div class="network-icon icon-instagram"></div>
            <shopspot-icon tile-id="undefined" />
        </div>
    </div>
  <img class='fallback-image' tile-id="undefined" src="undefined" loading="lazy" width="120"/>
  <video style="display:none;" reload="metadata" class="video-content" loading="lazy"  
  oncanplaythrough="this.removeAttribute('controls'); 
  this.muted=true; 
  this.parentElement.querySelector('.fallback-image').style.display='none'; 
  this.style.display='flex';
  this.parentElement.parentElement.parentElement.querySelectorAll('.icon-play').forEach((el) => {
    el.style.display='none';
  });
  " 
  controls autoplay muted loop>
  <source src="https://example.com/&lt;script&gt;alert(1)&lt;/script&gt;.mp4#t=0.1" type="video/mp4">
  </video>"
`;

exports[`loadPlayVideoHelper should generate a video tag with url, width and height 1`] = `
"<div class="icon-section">
        <div class="top-section">
            
            
        </div>
        <div class="center-section">
            
        </div>
        <div class="bottom-section">
            <div class="network-icon icon-instagram"></div>
            <shopspot-icon tile-id="undefined" />
        </div>
    </div>
  <img class='fallback-image' tile-id="undefined" src="undefined" loading="lazy" width="120"/>
  <video style="display:none;" reload="metadata" class="video-content" loading="lazy"  
  oncanplaythrough="this.removeAttribute('controls'); 
  this.muted=true; 
  this.parentElement.querySelector('.fallback-image').style.display='none'; 
  this.style.display='flex';
  this.parentElement.parentElement.parentElement.querySelectorAll('.icon-play').forEach((el) => {
    el.style.display='none';
  });
  " 
  controls autoplay muted loop>
  <source src="https://example.com/video.mp4#t=0.1" type="video/mp4">
  </video>"
`;

exports[`loadPlayVideoHelper should generate a video with embedly html 1`] = `
"<div class="icon-section">
        <div class="top-section">
            
            
        </div>
        <div class="center-section">
            
        </div>
        <div class="bottom-section">
            <div class="network-icon icon-tiktok"></div>
            <shopspot-icon tile-id="undefined" />
        </div>
    </div>
  <img class='fallback-image' tile-id="undefined" src="undefined" loading="lazy" width="120"/>
  <iframe loading="lazy" onload="this.removeAttribute('controls'); 
  this.muted=true; 
  this.parentElement.querySelector('.fallback-image').style.display='none'; 
  this.style.display='flex';
  this.parentElement.parentElement.parentElement.querySelectorAll('.icon-play').forEach((el) => {
    el.style.display='none';
  });
  " class="video-content" allowfullscreen src="https://www.tiktok.com/player/v1/7226405581324651819?autoplay=1&loop=1&controls=0"></iframe>"
`;

exports[`loadPlayVideoHelper should handle undefined width and height gracefully 1`] = `
"<div class="icon-section">
        <div class="top-section">
            
            
        </div>
        <div class="center-section">
            
        </div>
        <div class="bottom-section">
            <div class="network-icon icon-instagram"></div>
            <shopspot-icon tile-id="undefined" />
        </div>
    </div>
  <img class='fallback-image' tile-id="undefined" src="undefined" loading="lazy" width="120"/>
  <video style="display:none;" reload="metadata" class="video-content" loading="lazy"  
  oncanplaythrough="this.removeAttribute('controls'); 
  this.muted=true; 
  this.parentElement.querySelector('.fallback-image').style.display='none'; 
  this.style.display='flex';
  this.parentElement.parentElement.parentElement.querySelectorAll('.icon-play').forEach((el) => {
    el.style.display='none';
  });
  " 
  controls autoplay muted loop>
  <source src="https://example.com/video.mp4#t=0.1" type="video/mp4">
  </video>"
`;

exports[`loadPlayVideoHelper should return an empty string if URL is null or undefined 1`] = `
"<div class="icon-section">
        <div class="top-section">
            
            
        </div>
        <div class="center-section">
            
        </div>
        <div class="bottom-section">
            <div class="network-icon icon-instagram"></div>
            <shopspot-icon tile-id="undefined" />
        </div>
    </div>
  <img class='fallback-image' tile-id="undefined" src="undefined" loading="lazy" width="120"/>
  <video style="display:none;" reload="metadata" class="video-content" loading="lazy"  
  oncanplaythrough="this.removeAttribute('controls'); 
  this.muted=true; 
  this.parentElement.querySelector('.fallback-image').style.display='none'; 
  this.style.display='flex';
  this.parentElement.parentElement.parentElement.querySelectorAll('.icon-play').forEach((el) => {
    el.style.display='none';
  });
  " 
  controls autoplay muted loop>
  <source src="#t=0.1" type="video/mp4">
  </video>"
`;
